core/domain/draft_upgrade_services.py

```diff
@@ -0,0 +1,82 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Commands that can be used to upgrade draft to newer Exploration versions."""
+import logging
+
+from core.domain import exp_domain
+from core.platform import models
+
+(exp_models, feedback_models, user_models) = models.Registry.import_models([
+    models.NAMES.exploration, models.NAMES.feedback, models.NAMES.user
+])
+
+
+def try_upgrade_from_version_to_version(
+        draft_change_list, from_version, to_version, exp_id):
+    """Try upgrading an ExplorationChange domain object by calling convert
+    functions for a list of commits.
+
+    For now, this handles only the scenario where commits_list contains a
+    single commit that migrates the state schema.
+
+    Args:
+        draft_change_list: list(dict). Each dict can be parsed as an
+            ExplorationChange domain object.
+        from_version: int. Current draft version.
+        to_version: int. Target exploration version.
+        exp_id: string. Exploration id.
+    Returns:
+        draft_change_list: list(dict). Upgraded change list. None if upgrade
+        failed.
+    """
+    if from_version >= to_version:
+        return
+    exp_versions = range(from_version + 1, to_version + 1)
+    commits_list = [
+        exp_models.ExplorationCommitLogEntryModel.get_commit(
+            exp_id, exp_version) for exp_version in exp_versions]
+    upgrade_times = 0
+    while from_version + upgrade_times < to_version:
+        commit = commits_list[upgrade_times]
+        if (
+                len(commit.commit_cmds) == 1 and
+                commit.commit_cmds[0]['cmd'] ==
+                exp_domain.CMD_MIGRATE_STATES_SCHEMA_TO_LATEST_VERSION):
+
+            conversion_fn_name = '_convert_v%s_dict_to_v%s_dict' % (
+                (from_version + upgrade_times),
+                (from_version + upgrade_times + 1))
+            if not hasattr(DraftUpgradeUtil, conversion_fn_name):
+                logging.warning('%s is not implemented' % conversion_fn_name)
+                return None
+            conversion_fn = getattr(DraftUpgradeUtil, conversion_fn_name)
+            draft_change_list = conversion_fn(draft_change_list)
+            if not draft_change_list:
+                logging.warning(
+                    'Cannot upgrade due to %s is not defined' %
+                    conversion_fn_name)
+                return
+        else:
+            return
+        upgrade_times += 1
+    if from_version + upgrade_times >= to_version:
+        return draft_change_list
+
+
+class DraftUpgradeUtil(object):
```

Comment:
Hm, I worry a little bit that this introduces skew. It's meant for tests, but main code is depending heavily on it.

What do you think about the following strategy instead -- don't have this class, but in tests, swap out the CURRENT_EXPLORATION_SCHEMA_VERSION (and possibly similarly for state schema version) constant strategically to simulate an environment change? For example, start by swapping it out for v31 and creating an exploration. Then, swap it out for v32 -- now, you're in an environment where the schema version has been upgraded. That way, you don't need to introduce new stuff just for tests.

You can also choose the specific versions to use so that the upgrades done in the corresponding conversion methods in exp_domain.py are fairly straightforward.

Reply:
The intention for this class isn't just for tests, this class is here so that we can call getattr(DraftUpgradeUtil, conversion_function_name) to get the conversion function to invoke. If not defining a class, the only other place I thought of is to put these conversion functions in the draft_upgrade_services module itself, but it's not as clear how to get which conversion function to invoke.

Reply:
Ah, in that case, I think you can actually go ahead and populate this with conversion functions from the last 1-2 schema upgrades.

Reply:
Done.